File: imageviewer/src/main/java/com/stfalcon/imageviewer/common/gestures/dismiss/SwipeToDismissHandler.java
package com.stfalcon.imageviewer.common.gestures.dismiss;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.ObjectAnimator;
import android.annotation.SuppressLint;
import android.view.MotionEvent;
import android.view.View;
import android.view.View.OnTouchListener;
import android.view.animation.AccelerateInterpolator;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.stfalcon.imageviewer.common.BooleanSupplier;
import com.stfalcon.imageviewer.common.tools.Views;

public final class SwipeToDismissHandler implements OnTouchListener {
  private static final long ANIMATION_DURATION = 200L;

  private static final AccelerateInterpolator accelerateInterpolator = new AccelerateInterpolator();

  private final View swipeView;
  private final Runnable onDismiss;
  private final OnSwipeViewModed onSwipeViewMove;
  private final BooleanSupplier shouldAnimateDismiss;

  private int translationLimit;
  private boolean isTracking;
  private float startY;

  public SwipeToDismissHandler(@NonNull View swipeView, @NonNull Runnable onDismiss,
      @NonNull OnSwipeViewModed onSwipeViewMove, @NonNull BooleanSupplier shouldAnimateDismiss) {
    this.swipeView = swipeView;
    this.onDismiss = onDismiss;
    this.onSwipeViewMove = onSwipeViewMove;
    this.shouldAnimateDismiss = shouldAnimateDismiss;
    this.translationLimit = swipeView.getHeight() / 4;
  }

  @SuppressLint({"ClickableViewAccessibility"})
  @Override
  public boolean onTouch(@NonNull View v, @NonNull MotionEvent event) {
    switch (event.getAction()) {
      case MotionEvent.ACTION_DOWN:
        if (Views.getHitRect(swipeView).contains((int) event.getX(), (int) event.getY())) {
          isTracking = true;
        }

        startY = event.getY();
        return true;
      case MotionEvent.ACTION_UP:
      case MotionEvent.ACTION_CANCEL:
        if (isTracking) {
          isTracking = false;
          onTrackingEnd(v.getHeight());
        }

        return true;
      case MotionEvent.ACTION_MOVE:
        if (isTracking) {
          float translationY = event.getY() - startY;
          swipeView.setTranslationY(translationY);
          onSwipeViewMove.onSwipeViewModed(translationY, translationLimit);
        }

        return true;
      default:
        return false;
    }
  }

  public void initiateDismissToBottom() {
    animateTranslation((float) swipeView.getHeight(), ANIMATION_DURATION);
  }

  private void onTrackingEnd(int parentHeight) {
    float animateTo =
        swipeView.getTranslationY() < ((float) -translationLimit) ? -((float) parentHeight)
            : swipeView.getTranslationY() > (float) translationLimit ? (float) parentHeight
                : 0f;
    if (animateTo != 0f && !shouldAnimateDismiss.getAsBoolean()) {
      onDismiss.run();
    } else {
      animateTranslation(animateTo, ANIMATION_DURATION);
    }
  }

  private void animateTranslation(float translationTo, long duration) {
    ObjectAnimator animator = ObjectAnimator.ofFloat(swipeView, View.TRANSLATION_Y, translationTo)
        .setDuration(duration);
    animator.setInterpolator(accelerateInterpolator);
    animator.addUpdateListener(
        listener -> onSwipeViewMove.onSwipeViewModed(swipeView.getTranslationY(),
            translationLimit));
    animator.addListener(new AnimatorListenerAdapter() {
      public void onAnimationEnd(@Nullable Animator animation) {
        if (translationTo != 0f) {
          onDismiss.run();
        }
      }
    });
    animator.start();
  }
}

File: imageviewer/src/main/java/com/stfalcon/imageviewer/common/gestures/dismiss/SwipeToDismissHandler.kt
/*
 * Copyright 2018 stfalcon.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.stfalcon.imageviewer.common.gestures.dismiss

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.ObjectAnimator
import android.annotation.SuppressLint
import android.view.MotionEvent
import android.view.View
import android.view.animation.AccelerateInterpolator
import com.stfalcon.imageviewer.common.extensions.hitRect

internal class SwipeToDismissHandler(
    private val swipeView: View,
    private val onDismiss: Runnable,
    private val onSwipeViewMove: (translationY: Float, translationLimit: Int) -> Unit,
    private val shouldAnimateDismiss: () -> Boolean
) : View.OnTouchListener {

    companion object {
        private const val ANIMATION_DURATION = 200L
        private val accelerateInterpolator = AccelerateInterpolator()
    }

    private var translationLimit: Int = swipeView.height / 4
    private var isTracking = false
    private var startY: Float = 0f

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouch(v: View, event: MotionEvent): Boolean {
        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
                if (swipeView.hitRect.contains(event.x.toInt(), event.y.toInt())) {
                    isTracking = true
                }
                startY = event.y
                return true
            }
            MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> {
                if (isTracking) {
                    isTracking = false
                    onTrackingEnd(v.height)
                }
                return true
            }
            MotionEvent.ACTION_MOVE -> {
                if (isTracking) {
                    val translationY = event.y - startY
                    swipeView.translationY = translationY
                    onSwipeViewMove(translationY, translationLimit)
                }
                return true
            }
            else -> {
                return false
            }
        }
    }

    internal fun initiateDismissToBottom() {
        animateTranslation(swipeView.height.toFloat(), ANIMATION_DURATION)
    }

    private fun onTrackingEnd(parentHeight: Int) {
        val animateTo = when {
            swipeView.translationY < -translationLimit -> -parentHeight.toFloat()
            swipeView.translationY > translationLimit -> parentHeight.toFloat()
            else -> 0f
        }

        if (animateTo != 0f && !shouldAnimateDismiss()) {
            onDismiss.run()
        } else {
            animateTranslation(animateTo, ANIMATION_DURATION)
        }
    }

    private fun animateTranslation(translationTo: Float, duration: Long) {
        val animator = ObjectAnimator.ofFloat(swipeView, View.TRANSLATION_Y, translationTo)
            .setDuration(duration)
        animator.interpolator = accelerateInterpolator
        animator.addUpdateListener { onSwipeViewMove(swipeView.translationY, translationLimit) }
        animator.addListener(object : AnimatorListenerAdapter() {
            override fun onAnimationEnd(animation: Animator?) {
                if (translationTo != 0f) {
                    onDismiss.run()
                }
            }
        })
        animator.start()
    }
}